src/java_bytecode/java_bytecode_convert_method.cpp

```diff
@@ -21,15 +22,101 @@ Author: Daniel Kroening, kroening@kroening.com
 #include <goto-programs/cfg.h>
 #include <analyses/cfg_dominators.h>
 
+#include "bytecode_info.h"
 #include "java_bytecode_convert_method.h"
 #include "java_bytecode_convert_method_class.h"
-#include "bytecode_info.h"
+#include "java_object_factory.h"
 #include "java_types.h"
 
 #include <limits>
 #include <algorithm>
 #include <functional>
 #include <unordered_set>
+#include <regex>
+
+/*******************************************************************\
+
+Function: traverse_expr_tree
+
+  Inputs: `expr`: an expression tree to traverse
+          `parents`: will hold previously-visited nodes
+          `func`: will be called on each node, takes the node and the `parents`
+                  stack as arguments
+
+ Outputs: None
+
+ Purpose: Abstracts the process of calling a function on each node of the 
+          expression tree.
+
+\*******************************************************************/
+
+template <typename Func>
+static void traverse_expr_tree(
+  exprt &expr,
+  std::vector<exprt*> &parents,
+  Func func)
+{
+  const auto& parents_ref=parents;
+  func(expr, parents_ref);
+
+  parents.push_back(&expr);
+  for(auto &op : expr.operands())
+  {
+    traverse_expr_tree(op, parents, func);
+  }
+  parents.pop_back();
+}
+
+/*******************************************************************\
+
+Function: traverse_expr_tree
+
+  Inputs: `expr`: an expression tree
+
+ Outputs: None
+
+ Purpose: Looks for a symbol node with identifier 'nondetWith(out?)Null'.
+          Finds the following statement, checks if it is an "assert", and
+          replaces it with a code_skipt if so.
+
+\*******************************************************************/
+
+static void remove_assert_after_generic_nondet(exprt &expr)
+{
+  std::vector<exprt*> parents;
+  traverse_expr_tree(
+    expr,
+    parents,
+    [] (exprt &expr, const std::vector<exprt*>& parents)
+    {
+      const std::regex id_regex(
+        ".*org.cprover.CProver.(nondetWithNull|nondetWithoutNull).*");
+      if(expr.id()==ID_symbol &&
+         std::regex_match(as_string(to_symbol_expr(expr).get_identifier()),
+                          id_regex))
+      {
+        assert(2<=parents.size());
+        const auto before_1=*(parents.end()-1);
+        const auto before_2=*(parents.end()-2);
+
+        for(auto it=before_2->operands().begin(),
+                 end=before_2->operands().end();
+            it!=end;
+            ++it)
+        {
+          if(&(*it)==before_1)
+          {
+            assert(it+1!=end);
```

Comment:
This makes an assumption about the implementation of exprt::operandst. Use a second iterator and increment to point at the next element.